Modules for gathering facts about ZFS datasets and pools

xen0l opened the pull request:
##### ISSUE TYPE

<!--- Pick one below and delete the rest: -->
- New Module Pull Request
##### COMPONENT NAME

<!--- Name of the plugin/module/task -->

zfs_facts.py
zpool_facts.py
##### ANSIBLE VERSION

<!--- Paste verbatim output from “ansible --version” between quotes below -->

```
ansible 2.1.2.0
  config file = 
  configured module search path = Default w/o overrides
```
##### SUMMARY

<!--- Describe the change, including rationale and design decisions -->

These modules return facts about ZFS datasets and pools. They can be used for much better ZFS management with Ansible. They return ZFS dataset/pools properties as facts and based on those one can take decisions, e.g. delete snapshots older than X, report space usage, report property values, check ZFS pool health status etc.  These modules should run on all ZFS platforms and should work for both OpenZFS and ZFS implementations.

I also created a new category "zfs" under storage and move zfs.py there, so modules are kept together.

xen0l committed:
Add modules for gathering facts about ZFS datasets and pools

xen0l committed:
Move zfs module to storage/zfs subcategory

xen0l commented:
Migrated from https://github.com/ansible/ansible-modules-extras/pull/3075

xen0l commented:
@ansible, I got this reviewed. Can this be merged?

mattclay commented:
@xen0l The [tests](https://app.shippable.com/runs/584c7d7c3a097e0f00d7dcae/1/console) for this PR did not pass.

Try running `ansible-doc zfs zfs_facts zpool_facts` locally to see the failure.

xen0l commented:
@mattclay sorry for the issue. I fixed it. It now passes checks.

pyhalov commented on lib/ansible/modules/storage/zfs/zfs_facts.py:
```diff
@@ -0,0 +1,198 @@
+    recurse:
+        description:
+            - Specifies if properties for any children should be recursively
+              display.
```
should be displayed

xen0l replied:
Fixed.

pyhalov commented on lib/ansible/modules/storage/zfs/zfs_facts.py:
```diff
@@ -0,0 +1,198 @@
+
+        if rc == 0:
+            for line in out.splitlines():
+                dataset, property, value = line.split('\t')
```
Does line parsing depend on presence on self.parsable? BTW, why could we need non-parsable output?

xen0l replied:
No, parsable=True will return non-human friendly values (1K vs 1000) as _zfs get -p_ would.

pyhalov replied:
OK

xen0l committed:
Replace dict.iteritems() with iteritems(dict)

xen0l commented:
ready_for_review

xen0l commented:
@sivel you closed ansible/ansible-modules-extras#3075, which was reviewed there. This was reviewed by @jurajlutter there. Any chance to get this merged?

xen0l committed:
Add ANSIBLE_METADATA
Document return values
Make imports explicit
Use native YAML syntax in EXAMPLES

xen0l committed:
Add zfs_facts and zpool_facts modules to CHANGELOG.md

xen0l committed:
Add facts to return values
